external-crates/move/crates/move-compiler/src/linters/redundant_ref_deref.rs

```diff
@@ -0,0 +1,161 @@
+                //     }
+                //     UnannotatedExp_::Borrow(false, borrow_exp, field) => {
+                //         if self.stored_mutations.iter().any(|stored| {
+                //             Self::compare_exp(&stored.exp, &borrow_exp) && stored.field == *field
+                //         }) {
+                //             self.add_redundant_ref_deref_diagnostic(exp.exp.loc);
+                //         }
+                //     }
+                //     _ => {}
+                // };
+            }
+        }
+    }
+
+    fn check_borrow_pattern(&mut self, exp: &Exp, deref_exp: &Box<Exp>) {
+        match &deref_exp.exp.value {
+            UnannotatedExp_::BorrowLocal(false, _) | UnannotatedExp_::TempBorrow(false, _) => {
+                self.add_redundant_ref_deref_diagnostic(exp.exp.loc);
+            }
+            UnannotatedExp_::Borrow(false, borrow_exp, field) => {
+                if self.is_stored_mutation(borrow_exp, field) {
+                    self.add_redundant_ref_deref_diagnostic(exp.exp.loc);
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn is_stored_mutation(&self, borrow_exp: &Box<Exp>, field: &Field) -> bool {
+        self.stored_mutations
+            .iter()
+            .any(|stored| Self::compare_exp(&stored.exp, borrow_exp) && &stored.field == field)
+    }
+
+    fn add_redundant_ref_deref_diagnostic(&mut self, loc: Loc) {
+        let diag = diag!(
+            REDUNDANT_REF_DEREF_DIAG,
+            (
+                loc,
+                "Redundant borrow-dereference detected. Consider removing the borrow-dereference operation and using the expression directly."
+            )
+        );
+        self.env.add_diag(diag);
+    }
+
+    fn track_modifications(&mut self, exp: &T::Exp) {
+        if let UnannotatedExp_::Mutate(lhs, _) = &exp.exp.value {
+            self.mark_modified(lhs);
+        }
+    }
+
+    fn mark_modified(&mut self, exp: &T::Exp) {
+        if let UnannotatedExp_::Borrow(_, inner_exp, field) = &exp.exp.value {
+            self.stored_mutations.push(StoredMutation {
+                exp: inner_exp.clone(),
+                field: field.clone(),
+            });
+        }
+    }
+
+    fn compare_exp(exp1: &Box<Exp>, exp2: &Box<Exp>) -> bool {
+        match (&exp1.exp.value, &exp2.exp.value) {
+            (
+                UnannotatedExp_::Borrow(_, inner1, field1),
+                UnannotatedExp_::Borrow(_, inner2, field2),
+            ) => {
+                field1 == field2
+                    && (Self::compare_unannotated_exp(inner1, inner2)
+                        || Self::compare_exp(&inner1, &inner2))
+            }
+            _ => Self::compare_unannotated_exp(exp1, exp2),
+        }
+    }
+
+    fn compare_unannotated_exp(exp1: &Box<Exp>, exp2: &Box<Exp>) -> bool {
+        std::mem::discriminant(&exp1.exp.value) == std::mem::discriminant(&exp2.exp.value)
+            && exp1.ty == exp2.ty
+    }
+}
```

Comment:
`track_modifications` and `mark_modified`:

`track_modifications` is called for every expression during the AST traversal.
When it encounters a mutation (assignment) operation, it calls `mark_modified` on the left-hand side of the assignment.
`mark_modified` checks if the expression is a borrow. If so, it records this mutation by adding a StoredMutation to the `stored_mutations` vector.
This process builds up a record of all the mutations in the code, which will be used later to determine if certain ref/deref patterns are truly redundant.


`check_redundant_ref_deref` and `check_borrow_pattern`:

`check_redundant_ref_deref` is also called for every expression during the AST traversal.
It looks for a specific pattern: a temporary borrow of a dereference.
When it finds this pattern, it calls `check_borrow_pattern` with the full expression and the dereferenced expression.
`check_borrow_pattern` then examines the dereferenced expression more closely:

If it's a `BorrowLocal` or `TempBorrow`, it immediately flags it as redundant.
If it's a `Borrow`, it calls `is_stored_mutation` to check if this borrow has been mutated elsewhere in the code.

`is_stored_mutation`:

This function is called by `check_borrow_pattern` when it encounters a `Borrow` expression.
It searches through the `stored_mutations` vector (which was populated by `mark_modified`) to see if this particular borrow has been recorded as a mutation.
To compare borrows, it uses the `compare_exp` function.

`compare_exp` and `compare_unannotated_exp`:

`compare_exp` is used by `is_stored_mutation` to compare two expressions for equality.
It handles the special case of `Borrow` expressions, comparing their inner expressions and fields.
For all other cases, it delegates to `compare_unannotated_exp` .If where is another nested fields, recursive utils hit the last fields.
`compare_unannotated_exp` compares two expressions by checking if they have the same type and the same kind of expression (using std::mem::discriminant).

The recursion allows the function to handle nested Borrow expressions. Here's why this is necessary:

In Move, it's possible to have nested borrows, like`&*&x.y.z.g`.
When comparing two such nested borrow expressions, we need to traverse through all layers of borrowing to compare the innermost expressions.
The recursive call to compare_exp allows the function to "peel off" each layer of borrowing one at a time, continuing until it reaches non-Borrow expressions or finds a mismatch.